llap-tez/src/java/org/apache/hadoop/hive/llap/tezplugins/metrics/BlacklistingLlapMetricsListener.java

```diff
@@ -0,0 +1,200 @@
+  }
+
+  @Override
+  public void newDaemonMetrics(String workerIdentity, LlapMetricsCollector.LlapMetrics newMetrics) {
+    // no op
+  }
+
+  @Override
+  public void newClusterMetrics(Map<String, LlapMetricsCollector.LlapMetrics> newMetrics) {
+    long sumAverageTime = 0;
+    long sumEmptyExecutors = 0;
+    long maxAverageTime = 0;
+    long maxAverageTimeEmptyExecutors = 0;
+    long maxAverageTimeMaxExecutors = 0;
+    long workerNum = 0;
+    int blacklistedNodes = 0;
+    String maxAverageTimeIdentity = null;
+    for (String workerIdentity : newMetrics.keySet()) {
+      Map<String, Long> metrics = newMetrics.get(workerIdentity).getMetrics();
+      long requestHandled = metrics.get(LlapDaemonExecutorInfo.ExecutorTotalRequestsHandled.name());
+      long averageTime = metrics.get(LlapDaemonExecutorInfo.AverageResponseTime.name());
+      long emptyExecutor =
+          metrics.get(LlapDaemonExecutorInfo.ExecutorNumExecutorsAvailableAverage.name());
+      long maxExecutors = metrics.get(LlapDaemonExecutorInfo.ExecutorNumExecutorsPerInstance.name());
+
+      LOG.debug("Checking node {} with data: " +
+                    "requestHandled={}, " +
+                    "averageTime={}, " +
+                    "emptyExecutors={}, " +
+                    "maxExecutors={}",
+          workerIdentity, requestHandled, averageTime, emptyExecutor, maxExecutors);
+
+      if (maxExecutors == 0) {
+        blacklistedNodes++;
+        if (blacklistedNodes >= this.maxBlacklistedNodes) {
+          LOG.debug("Already too many blacklisted nodes. Skipping.");
+          return;
+        }
+      }
+
+      if (requestHandled > this.minServedTasksNumber) {
+        workerNum++;
+        sumAverageTime += averageTime;
+        if ( averageTime > maxAverageTime) {
+          maxAverageTime = averageTime;
+          maxAverageTimeEmptyExecutors = emptyExecutor;
+          maxAverageTimeMaxExecutors = maxExecutors;
+          maxAverageTimeIdentity = workerIdentity;
+        }
+        sumEmptyExecutors += emptyExecutor;
+      }
+    }
+
+    // If we do not have enough data then return.
+    if (workerNum < 2) {
+      return;
+    }
+
+    LOG.debug("Found slowest node {} with data: " +
+                  "sumAverageTime={}, " +
+                  "sumEmptyExecutors={}, " +
+                  "maxAverageTime={}, " +
+                  "maxAverageTimeEmptyExecutors={}, " +
+                  "maxAverageTimeMaxExecutors={}, " +
+                  "workerNum={}, " +
+                  "maxAverageTimeIdentity={}, " +
+                  "blacklistedNodes={}",
+        sumAverageTime, sumEmptyExecutors, maxAverageTime, maxAverageTimeEmptyExecutors,
+        maxAverageTimeMaxExecutors, workerNum, maxAverageTimeIdentity, blacklistedNodes);
+    // Check if the slowest node is at least timeThreshold times slower than the average
+    long averageTimeWithoutSlowest = (sumAverageTime - maxAverageTime) / (workerNum - 1);
```

Comment:
I would already create a double here to avoid loss of precision. (case sumAverageTime - maxAverageTime to double before division).

Reply:
It should be a very specific case where it can cause problems. But done it anyway, since it does not hurt